File: backend/routes/expenses.js
const express = require('express');
const router  = express.Router();
const { v4: uuidv4 } = require('uuid');
const db   = require('../database');
const auth = require('../middleware/auth');

// GET /api/expenses — all expenses for current user's groups
router.get('/', auth, (req, res) => {
  const { group_id, category, limit = 50 } = req.query;

  let query = `
    SELECT e.*, u.name AS paid_by_name, g.name AS group_name, g.emoji AS group_emoji
    FROM expenses e
    JOIN users   u ON u.id = e.paid_by
    JOIN groups  g ON g.id = e.group_id
    JOIN group_members gm ON gm.group_id = g.id AND gm.user_id = ?
  `;
  const params = [req.userId];

  if (group_id) { query += ' AND e.group_id = ?'; params.push(group_id); }
  if (category)  { query += ' AND e.category = ?'; params.push(category); }
  query += ' ORDER BY e.created_at DESC LIMIT ?';
  params.push(Number(limit));

  const expenses = db.prepare(query).all(...params);
  res.json(expenses);
});

// POST /api/expenses — create expense
router.post('/', auth, (req, res) => {
  const { group_id, title, amount, category = 'other', split_type = 'equal', emoji = '💸', splits } = req.body;
  if (!group_id || !title || !amount) return res.status(400).json({ error: 'group_id, title, amount required' });

  const isMember = db.prepare('SELECT 1 FROM group_members WHERE group_id=? AND user_id=?').get(group_id, req.userId);
  if (!isMember) return res.status(403).json({ error: 'Not a member of this group' });

  const id = uuidv4();
  db.prepare('INSERT INTO expenses (id,group_id,title,amount,category,paid_by,split_type,emoji) VALUES (?,?,?,?,?,?,?,?)')
    .run(id, group_id, title, Number(amount), category, req.userId, split_type, emoji);

  // Create splits
  const members = db.prepare('SELECT user_id FROM group_members WHERE group_id=?').all(group_id).map(m => m.user_id);

  if (split_type === 'equal' || !splits) {
    const share = Math.round((Number(amount) / members.length) * 100) / 100;
    const insertSplit = db.prepare('INSERT INTO expense_splits (id,expense_id,user_id,amount,is_settled) VALUES (?,?,?,?,?)');
    members.forEach(uid => {
      insertSplit.run(uuidv4(), id, uid, share, uid === req.userId ? 1 : 0);
    });
  } else {
    // Custom splits
    const insertSplit = db.prepare('INSERT INTO expense_splits (id,expense_id,user_id,amount,is_settled) VALUES (?,?,?,?,?)');
    Object.entries(splits).forEach(([uid, amt]) => {
      insertSplit.run(uuidv4(), id, uid, Number(amt), uid === req.userId ? 1 : 0);
    });
  }

  // Notify group members
  const insertNotif = db.prepare('INSERT INTO notifications (id,user_id,title,body,type) VALUES (?,?,?,?,?)');
  members.filter(uid => uid !== req.userId).forEach(uid => {
    insertNotif.run(uuidv4(), uid, 'New Expense Added',
      `New expense "${title}" of ₹${amount} was added to your group.`, 'info');
      
    if (req.io) {
      req.io.to(`user_${uid}`).emit('notification', {
        title: 'New Expense Added',
        body: `New expense "${title}" of ₹${amount} was added to your group.`
      });
      req.io.to(`user_${uid}`).emit('balance_update', { group_id });
    }
  });

  const expense = db.prepare(`SELECT e.*,u.name AS paid_by_name FROM expenses e JOIN users u ON u.id=e.paid_by WHERE e.id=?`).get(id);
  res.status(201).json(expense);
});

// GET /api/expenses/:id
router.get('/:id', auth, (req, res) => {
  const expense = db.prepare(`
    SELECT e.*, u.name AS paid_by_name, g.name AS group_name
    FROM expenses e JOIN users u ON u.id=e.paid_by JOIN groups g ON g.id=e.group_id
    WHERE e.id=?
  `).get(req.params.id);
  if (!expense) return res.status(404).json({ error: 'Expense not found' });

  const isMember = db.prepare('SELECT 1 FROM group_members WHERE group_id=? AND user_id=?').get(expense.group_id, req.userId);
  if (!isMember) return res.status(403).json({ error: 'Not a member' });

  const splits = db.prepare(`
    SELECT es.*, u.name FROM expense_splits es JOIN users u ON u.id=es.user_id WHERE es.expense_id=?
  `).all(expense.id);

  res.json({ ...expense, splits });
});

// PUT /api/expenses/:id
router.put('/:id', auth, (req, res) => {
  const expense = db.prepare('SELECT * FROM expenses WHERE id=?').get(req.params.id);
  if (!expense) return res.status(404).json({ error: 'not found' });
  if (expense.paid_by !== req.userId) return res.status(403).json({ error: 'Only the payer can edit' });

  const { title, amount, category, emoji } = req.body;
  db.prepare('UPDATE expenses SET title=COALESCE(?,title), amount=COALESCE(?,amount), category=COALESCE(?,category), emoji=COALESCE(?,emoji) WHERE id=?')
    .run(title, amount, category, emoji, expense.id);

  res.json(db.prepare('SELECT * FROM expenses WHERE id=?').get(expense.id));
});

// DELETE /api/expenses/:id
router.delete('/:id', auth, (req, res) => {
  const expense = db.prepare('SELECT * FROM expenses WHERE id=?').get(req.params.id);
  if (!expense) return res.status(404).json({ error: 'not found' });
  if (expense.paid_by !== req.userId) return res.status(403).json({ error: 'Only the payer can delete' });

  db.prepare('DELETE FROM expenses WHERE id=?').run(expense.id);
  res.json({ message: 'Expense deleted' });
});

module.exports = router;

File: backend/routes/expenses.py
"""
Splito Expenses Routes
GET    /api/expenses                   — my recent expenses
POST   /api/expenses                   — add expense
GET    /api/expenses/<id>              — detail
DELETE /api/expenses/<id>              — delete
PATCH  /api/expenses/<id>/settle       — settle my split
GET    /api/expenses/group/<group_id>  — expenses for a group
"""

from flask import Blueprint, request, jsonify
from database import get_conn, row_to_dict, rows_to_list, _uuid, _now
from utils.auth import require_auth

expenses_bp = Blueprint('expenses', __name__)


def _is_member(user_id, group_id):
    with get_conn() as conn:
        return conn.execute(
            "SELECT 1 FROM group_members WHERE group_id=? AND user_id=?",
            (group_id, user_id)
        ).fetchone()


def _expense_with_splits(eid):
    with get_conn() as conn:
        e = row_to_dict(conn.execute("""
            SELECT e.*, u.username as paid_by_name
            FROM expenses e JOIN users u ON u.id = e.paid_by_id
            WHERE e.id=?
        """, (eid,)).fetchone())
        if not e:
            return None
        splits = rows_to_list(conn.execute("""
            SELECT es.*, u.username
            FROM expense_splits es JOIN users u ON u.id = es.user_id
            WHERE es.expense_id=?
        """, (eid,)).fetchall())
        e['splits'] = splits
        return e


@expenses_bp.route('', methods=['GET'])
@require_auth
def list_my_expenses(current_user):
    uid = current_user['id']
    with get_conn() as conn:
        rows = conn.execute("""
            SELECT e.*, u.username as paid_by_name
            FROM expenses e
            JOIN users u ON u.id = e.paid_by_id
            JOIN group_members gm ON gm.group_id = e.group_id
            WHERE gm.user_id=?
            ORDER BY e.created_at DESC LIMIT 50
        """, (uid,)).fetchall()
    return jsonify({'expenses': rows_to_list(rows)})


@expenses_bp.route('', methods=['POST'])
@require_auth
def add_expense(current_user):
    data         = request.get_json(silent=True) or {}
    group_id     = data.get('group_id')
    title        = (data.get('title') or '').strip()
    amount       = data.get('amount')
    category     = data.get('category', 'Other')
    split_type   = data.get('split_type', 'equal')
    participants = data.get('participants', [])
    custom_splits= data.get('custom_splits', {})
    notes        = data.get('notes', '')

    if not all([group_id, title, amount, participants]):
        return jsonify({'error': 'group_id, title, amount, participants are required'}), 400
    try:
        amount = float(amount)
        if amount <= 0: raise ValueError
    except (ValueError, TypeError):
        return jsonify({'error': 'amount must be a positive number'}), 400

    if not _is_member(current_user['id'], group_id):
        return jsonify({'error': 'Not a member of this group'}), 403

    # Build split map
    if split_type == 'equal':
        per = round(amount / len(participants), 2)
        splits_map = {uid: per for uid in participants}
    elif split_type == 'unequal':
        if not custom_splits:
            return jsonify({'error': 'custom_splits required for unequal split'}), 400
        total_c = sum(float(v) for v in custom_splits.values())
        if abs(total_c - amount) > 0.10:
            return jsonify({'error': f'custom_splits sum ({total_c:.2f}) must equal amount ({amount:.2f})'}), 400
        splits_map = {uid: float(v) for uid, v in custom_splits.items()}
    else:
        return jsonify({'error': 'split_type must be equal or unequal'}), 400

    eid = _uuid()
    with get_conn() as conn:
        conn.execute("INSERT INTO expenses VALUES (?,?,?,?,?,?,?,?,?,?)",
            (eid, group_id, current_user['id'], title, amount,
             category, split_type, notes, '', _now()))
        for uid, owed in splits_map.items():
            conn.execute("INSERT INTO expense_splits VALUES (?,?,?,?,?,?)",
                (_uuid(), eid, uid, owed, 0, None))
        conn.commit()

    return jsonify({'expense': _expense_with_splits(eid)}), 201


@expenses_bp.route('/group/<group_id>', methods=['GET'])
@require_auth
def group_expenses(current_user, group_id):
    if not _is_member(current_user['id'], group_id):
        return jsonify({'error': 'Not a member'}), 403

    page     = int(request.args.get('page', 1))
    per_page = int(request.args.get('per_page', 20))
    offset   = (page - 1) * per_page

    with get_conn() as conn:
        total = conn.execute(
            "SELECT COUNT(*) as c FROM expenses WHERE group_id=?", (group_id,)
        ).fetchone()['c']
        rows = conn.execute("""
            SELECT e.*, u.username as paid_by_name
            FROM expenses e JOIN users u ON u.id = e.paid_by_id
            WHERE e.group_id=? ORDER BY e.created_at DESC
            LIMIT ? OFFSET ?
        """, (group_id, per_page, offset)).fetchall()

    return jsonify({
        'expenses': rows_to_list(rows),
        'total':    total,
        'page':     page,
        'pages':    (total + per_page - 1) // per_page,
    })


@expenses_bp.route('/<expense_id>', methods=['GET'])
@require_auth
def get_expense(current_user, expense_id):
    e = _expense_with_splits(expense_id)
    if not e:
        return jsonify({'error': 'Expense not found'}), 404
    if not _is_member(current_user['id'], e['group_id']):
        return jsonify({'error': 'Not a member'}), 403
    return jsonify({'expense': e})


@expenses_bp.route('/<expense_id>', methods=['DELETE'])
@require_auth
def delete_expense(current_user, expense_id):
    with get_conn() as conn:
        e = row_to_dict(conn.execute("SELECT * FROM expenses WHERE id=?", (expense_id,)).fetchone())
        if not e:
            return jsonify({'error': 'Expense not found'}), 404

        if e['paid_by_id'] != current_user['id']:
            gm = conn.execute(
                "SELECT role FROM group_members WHERE group_id=? AND user_id=?",
                (e['group_id'], current_user['id'])
            ).fetchone()
            if not gm or gm['role'] != 'admin':
                return jsonify({'error': 'Only the payer or group admin can delete this'}), 403

        conn.execute("DELETE FROM expense_splits WHERE expense_id=?", (expense_id,))
        conn.execute("DELETE FROM expenses WHERE id=?", (expense_id,))
        conn.commit()

    return jsonify({'message': 'Expense deleted'})


@expenses_bp.route('/<expense_id>/settle', methods=['PATCH'])
@require_auth
def settle_split(current_user, expense_id):
    with get_conn() as conn:
        split = row_to_dict(conn.execute(
            "SELECT * FROM expense_splits WHERE expense_id=? AND user_id=?",
            (expense_id, current_user['id'])
        ).fetchone())
        if not split:
            return jsonify({'error': 'No split found for this user'}), 404
        conn.execute(
            "UPDATE expense_splits SET is_settled=1, settled_at=? WHERE id=?",
            (_now(), split['id'])
        )
        conn.commit()
        split['is_settled'] = 1
        split['settled_at'] = _now()
    return jsonify({'split': split})
